swift/ql/test/query-tests/Security/CWE-089/sqlite3_c_api.swift

```diff
@@ -0,0 +1,163 @@
+		_ db: OpaquePointer?,
+		_ zSql: UnsafeRawPointer?,
+		_ nByte: Int32,
+		_ ppStmt: UnsafeMutablePointer<OpaquePointer?>?,
+		_ pzTail: UnsafeMutablePointer<UnsafeRawPointer?>?
+	) -> Int32 { return SQLITE_OK }
+
+func sqlite3_prepare16_v3(
+		_ db: OpaquePointer?,
+		_ zSql: UnsafeRawPointer?,
+		_ nByte: Int32,
+		_ prepFlags: UInt32,
+		_ ppStmt: UnsafeMutablePointer<OpaquePointer?>?,
+		_ pzTail: UnsafeMutablePointer<UnsafeRawPointer?>?
+	) -> Int32 { return SQLITE_OK }
+
+func sqlite3_bind_text(
+		_ _: OpaquePointer?,
+		_ _: Int32,
+		_ _: UnsafePointer<CChar>?,
+		_ _: Int32,
+		_ callback: (@convention(c) (UnsafeMutableRawPointer?) -> Void)?
+	) -> Int32 { return SQLITE_OK }
+// (many other `sqlite3_bind_*` functions exist)
+
+func sqlite3_step(
+		_ _: OpaquePointer?
+	) -> Int32 { return SQLITE_OK }
+
+func sqlite3_finalize(
+		_ pStmt: OpaquePointer?
+	) -> Int32 { return SQLITE_OK }
+
+func sanitize(_ string: String) -> String { return string }
+
+// --- tests ---
+
+func test_sqlite3_c_api(db: OpaquePointer?) {
+	let localString = "user"
+	let remoteString = try! String(contentsOf: URL(string: "http://example.com/")!)
+	let sanitizedString = sanitize(remoteString)
+
+	let unsafeQuery1 = remoteString
+	let unsafeQuery2 = "SELECT * FROM users WHERE username='" + remoteString + "'"
+	let unsafeQuery3 = "SELECT * FROM users WHERE username='\(remoteString)'"
+	let safeQuery1 = "SELECT * FROM users WHERE username='\(localString)'"
+	let safeQuery2 = "SELECT * FROM users WHERE username='\(sanitizedString)'"
+	let varQuery = "SELECT * FROM users WHERE username=?"
+
+	// --- exec ---
+
+	let result1 = sqlite3_exec(db, unsafeQuery1, nil, nil, nil) // BAD
+	let result2 = sqlite3_exec(db, unsafeQuery2, nil, nil, nil) // BAD
+	let result3 = sqlite3_exec(db, unsafeQuery3, nil, nil, nil) // BAD
+	let result4 = sqlite3_exec(db, safeQuery1, nil, nil, nil) // GOOD
+	let result5 = sqlite3_exec(db, safeQuery2, nil, nil, nil) // GOOD (sanitized)
+
+	// --- prepared statements ---
+
+	var stmt1: OpaquePointer?
+
+	if (sqlite3_prepare(db, unsafeQuery3, -1, &stmt1, nil) == SQLITE_OK) { // BAD
+		let result = sqlite3_step(stmt1)
+		// ...
+	}
+	sqlite3_finalize(stmt1)
+
+	var stmt2: OpaquePointer?
+
+	if (sqlite3_prepare(db, varQuery, -1, &stmt2, nil) == SQLITE_OK) { // GOOD
+		if (sqlite3_bind_text(stmt2, 1, localString, -1, SQLITE_TRANSIENT) == SQLITE_OK) { // GOOD
+			let result = sqlite3_step(stmt2)
+			// ...
+		}
+	}
+	sqlite3_finalize(stmt2)
+
+	var stmt3: OpaquePointer?
+
+	if (sqlite3_prepare(db, varQuery, -1, &stmt3, nil) == SQLITE_OK) { // GOOD
+		if (sqlite3_bind_text(stmt3, 1, sanitizedString, -1, SQLITE_TRANSIENT) == SQLITE_OK) { // GOOD
+			let result = sqlite3_step(stmt3)
+			// ...
+		}
+	}
+	sqlite3_finalize(stmt3)
+
+	var stmt4: OpaquePointer?
+
+	if (sqlite3_prepare_v2(db, varQuery, -1, &stmt4, nil) == SQLITE_OK) { // GOOD
+		if (sqlite3_bind_text(stmt4, 1, remoteString, -1, SQLITE_TRANSIENT) == SQLITE_OK) { // GOOD???
```

Comment:
This is safe indeed. That's the proper way of building SQL statements with untrusted data: using prepared statements and binding the untrusted data to the statement with the appropriate type.

Reply:
Thanks, good to confirm this is the right way.